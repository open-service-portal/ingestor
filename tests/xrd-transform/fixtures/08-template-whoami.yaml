# XRD: WhoAmIApp (XR Definition)
# Purpose: Define a simple demo application with automatic domain configuration
# Restaurant Analogy: The "menu item" - what customers can order
apiVersion: apiextensions.crossplane.io/v2
kind: CompositeResourceDefinition
metadata:
  name: whoamiapps.openportal.dev
  labels:
    terasky.backstage.io/generate-form: "true"
    openportal.dev/version: "dev"  # CI/CD will replace with actual version
  annotations:
    # Standard Backstage metadata
    backstage.io/title: "Who Am I App"
    backstage.io/description: "Simple demo application with automatic domain configuration"
    backstage.io/owner: "platform-team"
    backstage.io/system: "demo-applications"
    backstage.io/lifecycle: "production"
    backstage.io/source-location: "url:https://github.com/open-service-portal/template-whoami"

    # Project metadata
    openportal.dev/tags: "demo,application"
    openportal.dev/version: "dev"

    # Crossplane
    crossplane.io/version: "v2.0"

    # GitHub
    github.com/project-slug: "open-service-portal/template-whoami"
spec:
  scope: Namespaced  # Crossplane v2 - XRs can be created in any namespace
  group: openportal.dev
  names:
    kind: WhoAmIApp
    plural: whoamiapps
  # Default composition when no selector is specified
  defaultCompositionRef:
    name: whoamiapp
  versions:
  - name: v1alpha1
    served: true
    referenceable: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              name:
                type: string
                description: Application name (becomes subdomain, e.g., myapp.openportal.dev)
                pattern: "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
                minLength: 1
                maxLength: 63
              replicas:
                type: integer
                description: Number of replicas
                default: 1
                minimum: 1
                maximum: 3
              image:
                type: string
                description: Container image to deploy
                default: "traefik/whoami:v1.10.1"
            required: []
          status:
            type: object
            properties:
              domain:
                type: string
                description: The configured domain for this deployment
              ready:
                type: boolean
                description: Whether the application is ready