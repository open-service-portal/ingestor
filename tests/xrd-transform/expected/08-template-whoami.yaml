# ================================================================================================
# EXPECTED TEST OUTPUT - DO NOT BLINDLY REGENERATE
# ================================================================================================
# This file defines the expected output for xrd-transform tests.
#
# ‚ö†Ô∏è  IMPORTANT: Only update this file when:
#   1. You've added NEW FEATURES that intentionally change output
#   2. You've FIXED BUGS where the old output was incorrect
#   3. You've MODIFIED TEMPLATES with a clear purpose
#
# üö´ DO NOT simply copy from output/ without understanding the changes!
#
# Before updating:
#   - Run: bash tests/xrd-transform/run-tests.sh
#   - Review ALL differences carefully
#   - Understand WHY each change occurred
#   - Ensure both Template AND API entities are present
#   - Document the reason in your commit message
#
# This header is validated by the test runner.
# ================================================================================================

apiVersion: scaffolder.backstage.io/v1beta3
kind: Template
metadata:
  name: whoamiapps-openportal-dev
  title: Who Am I App
  description: Create and manage WhoAmIApp resources
  tags:
    - crossplane
    - openportal-dev
    - demo
    - application
  annotations:
    backstage.io/managed-by: xrd-transform
    backstage.io/managed-by-location: provider:XRDTemplateEntityProvider
    backstage.io/managed-by-origin-location: provider:XRDTemplateEntityProvider
    crossplane.io/xrd-name: whoamiapps.openportal.dev
    crossplane.io/xrd-group: openportal.dev
spec:
  owner: platform-team
  type: crossplane-resource
  parameters:
    - title: Resource Configuration
      required:
        - name
        - namespace
      properties:
        name:
          title: Name
          type: string
          description: Name of the WhoAmIApp
          pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
        namespace:
          title: Namespace
          type: string
          description: Kubernetes namespace
          default: default
        replicas:
          title: Replicas
          type: number
          description: Number of replicas
          default: 1
          minimum: 1
          maximum: 3
        image:
          title: Image
          type: string
          description: Container image to deploy
          default: null
  steps:
    - id: workflow-info
      name: Direct Apply Workflow Information
      action: debug:log
      input:
        message: |
          ‚ö° This template uses direct apply workflow:

          1. Creates your WhoAmIApp resource immediately in the cluster
          2. Registers the resource in the Backstage catalog

          Your resource will be created as soon as you click Create.
      output:
        text:
          - title: ‚ö° Direct Apply Workflow
            content: >
              This template follows a **direct apply workflow** for immediate
              resource creation:


              **What happens when you click Create:**


              1. **Create Resource Immediately** üöÄ
                 - Your WhoAmIApp resource is created directly in the Kubernetes cluster
                 - No approval required - resource is active immediately
                 - Applied to cluster: `rancher-desktop`
                 - Created in namespace: `$`

              2. **Register in Catalog** üìù
                 - The resource is automatically registered in the Backstage software catalog
                 - You can track and manage it through Backstage

              **‚úÖ Important**: Your resource is **created immediately** without
              approval workflow.
    - id: create-resource
      name: Create WhoAmIApp
      action: kube:apply
      input:
        manifest: |
          apiVersion: openportal.dev/v1alpha1
          kind: WhoAmIApp
          metadata:
            name: ${{ parameters.name }}
            namespace: ${{ parameters.namespace }}
          spec: ${{ parameters }}
        namespaced: true
      output:
        text:
          - title: Resource Details
            content: |
              **Resource Created Successfully!** ‚úÖ

              **Resource Information:**
              - **Name**: `$`
              - **Type**: WhoAmIApp
              - **API Version**: openportal.dev/v1alpha1
              - **Cluster**: rancher-desktop
              - **Namespace**: `$`

              Your resource is now active in the cluster and ready to use.
    - id: register
      name: Register in Catalog
      action: catalog:register
      input:
        repoContentsUrl: ${{ steps.create-resource.output.repoContentsUrl }}
        catalogInfoPath: /catalog-info.yaml
      output:
        links:
          - title: View in Catalog
            icon: catalog
            url: ${{ steps['register'].output.entityRef }}
        text:
          - title: Catalog Registration
            content: |
              **Registered in Backstage Catalog** üìù

              Your WhoAmIApp resource has been added to the software catalog.
              You can now track, monitor, and manage it through Backstage.

---
apiVersion: backstage.io/v1alpha1
kind: API
metadata:
  name: whoamiapps-openportal-dev-api
  title: Who Am I App API
  description: API specification for WhoAmIApp
  tags:
    - crossplane
    - ingestor
    - openportal-dev
  annotations:
    backstage.io/managed-by: xrd-transform
    backstage.io/managed-by-location: provider:XRDTemplateEntityProvider
    backstage.io/managed-by-origin-location: provider:XRDTemplateEntityProvider
    crossplane.io/xrd-name: whoamiapps.openportal.dev
spec:
  type: openapi
  lifecycle: production
  owner: platform-team
  definition: |
    openapi: 3.0.0
    info:
      title: WhoAmIApp API
      version: v1alpha1
      description: OpenAPI specification for WhoAmIApp
    paths:
      /whoamiapps:
        get:
          summary: List whoamiapps
          responses:
            '200':
              description: Success
        post:
          summary: Create WhoAmIApp
          requestBody:
            content:
              application/json:
                schema:
                  type: object
                  properties:
                    spec:
                      type: object
                      description: WhoAmIApp configuration
                      properties:
                        name:
                          type: string
                          description: "Application name (becomes subdomain, e.g., myapp.openportal.dev)"
                          pattern: ^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
                          minLength: 1
                          maxLength: 63
                        replicas:
                          type: number
                          description: "Number of replicas"
                          minimum: 1
                          maximum: 3
                        image:
                          type: string
                          description: "Container image to deploy"
                    status:
                      type: object
                      properties:
                        ready:
                          type: boolean
                          description: Whether the resource is ready
                        phase:
                          type: string
                          description: Current phase of the resource
                        message:
                          type: string
                          description: Status message
          responses:
            '201':
              description: Created

